source/tutorials/file-sets.md

```diff
@@ -0,0 +1,725 @@
+The [`difference`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.difference) function subtracts one file set from another.
+The result is a new file set that contains all files from the first argument that aren't in the second argument.
+
+Use it to filter out `./result` by changing the `sourceFiles` definition:
+
+```{code-block} diff
+:caption: build.nix
+ { stdenv, lib }:
+ let
+   fs = lib.fileset;
+-  sourceFiles = ./.;
++  sourceFiles = fs.difference ./. ./result;
+ in
+```
+
+Building this, the file set library will specify which files are taken from the directory:
+
+```shell-session
+$ nix-build
+trace: /home/user/fileset
+trace: - build.nix (regular)
+trace: - default.nix (regular)
+trace: - hello.txt (regular)
+trace: - nix (all files in directory)
+trace: - world.txt (regular)
+this derivation will be built:
+  /nix/store/zr19bv51085zz005yk7pw4s9sglmafvn-fileset.drv
+...
+'hello.txt' -> '/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset/hello.txt'
+'world.txt' -> '/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset/world.txt'
+...
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+An attempt to repeat the build will re-use the existing store path:
+
+```
+$ nix-build
+trace: /home/user/fileset
+trace: - build.nix (regular)
+trace: - default.nix (regular)
+trace: - hello.txt (regular)
+trace: - nix (all files in directory)
+trace: - world.txt (regular)
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+## Missing files
+
+Removing the `./result` symlink creates a new problem, though:
+
+```shell-session
+$ rm result
+$ nix-build
+error: lib.fileset.difference: Second argument (negative set)
+  (/home/user/select/result) is a path that does not exist.
+  To create a file set from a path that may not exist, use `lib.fileset.maybeMissing`.
+```
+
+Follow the instructions in the error message, and use `maybeMissing` to create a file set from a path that may not exist (in which case the file set will be empty):
+
+<!-- https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.maybeMissing -->
+
+```{code-block} diff
+:caption: build.nix
+ { stdenv, lib }:
+ let
+   fs = lib.fileset;
+-  sourceFiles = fs.difference ./. ./result;
++  sourceFiles = fs.difference ./. (fs.maybeMissing ./result);
+ in
+```
+
+This now works, using the whole directory since `./result` is not present:
+
+```
+$ nix-build
+trace: /home/user/select (all files in directory)
+this derivation will be built:
+  /nix/store/zr19bv51085zz005yk7pw4s9sglmafvn-fileset.drv
+...
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+Another build attempt will produce a different trace, but the same output path:
+
+```
+$ nix-build
+trace: /home/user/fileset
+trace: - build.nix (regular)
+trace: - default.nix (regular)
+trace: - hello.txt (regular)
+trace: - nix (all files in directory)
+trace: - world.txt (regular)
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+## Union (explicitly exclude files)
```

Comment:
```suggestion
## Union for explicit sets
```

Inclusion or exclusion is a matter of how you use it, but union is great for defining sets by explicit enumeration.

Reply:
Oh I get it now. You also have a section for explicit inclusion.
Probably still best to start with that.

Reply:
That would require rewriting it quite a lot, I suggest making that a follow-up, if at all.

Reply:
Yeah, follow up sounds better!